app/lib/.server/llm/create-summary.ts

```diff
@@ -0,0 +1,138 @@
+import type { IProviderSetting } from '~/types/model';
+import { DEFAULT_MODEL, DEFAULT_PROVIDER, PROVIDER_LIST } from '~/utils/constants';
+import { extractCurrentContext, extractPropertiesFromMessage, simplifyBoltActions } from './utils';
+import { createScopedLogger } from '~/utils/logger';
+import { LLMManager } from '~/lib/modules/llm/manager';
+
+const logger = createScopedLogger('create-summary');
+
+export async function createSummary(props: {
+  messages: Message[];
+  env: Env;
+  apiKeys?: Record<string, string>;
+  providerSettings?: Record<string, IProviderSetting>;
+  promptId?: string;
+  contextOptimization?: boolean;
+  onFinish?: (resp: GenerateTextResult<Record<string, CoreTool<any, any>>, never>) => void;
+}) {
+  const { messages, env: serverEnv, apiKeys, providerSettings, contextOptimization, onFinish } = props;
+  let currentModel = DEFAULT_MODEL;
+  let currentProvider = DEFAULT_PROVIDER.name;
+  const processedMessages = messages.map((message) => {
+    if (message.role === 'user') {
+      const { model, provider, content } = extractPropertiesFromMessage(message);
+      currentModel = model;
+      currentProvider = provider;
+
+      return { ...message, content };
+    } else if (message.role == 'assistant') {
+      let content = message.content;
+
+      if (contextOptimization) {
+        content = simplifyBoltActions(content);
+      }
+
+      return { ...message, content };
+    }
+
+    return message;
+  });
+
+  const provider = PROVIDER_LIST.find((p) => p.name === currentProvider) || DEFAULT_PROVIDER;
+  const staticModels = LLMManager.getInstance().getStaticModelListFromProvider(provider);
+  let modelDetails = staticModels.find((m) => m.name === currentModel);
+
+  if (!modelDetails) {
+    const modelsList = [
+      ...(provider.staticModels || []),
+      ...(await LLMManager.getInstance().getModelListFromProvider(provider, {
+        apiKeys,
+        providerSettings,
+        serverEnv: serverEnv as any,
+      })),
+    ];
+
+    if (!modelsList.length) {
+      throw new Error(`No models found for provider ${provider.name}`);
+    }
+
+    modelDetails = modelsList.find((m) => m.name === currentModel);
+
+    if (!modelDetails) {
+      // Fallback to first model
+      logger.warn(
+        `MODEL [${currentModel}] not found in provider [${provider.name}]. Falling back to first model. ${modelsList[0].name}`,
+      );
+      modelDetails = modelsList[0];
+    }
+  }
+
+  let slicedMessages = processedMessages;
+  const { summary } = extractCurrentContext(processedMessages);
+  let summaryText: string | undefined = undefined;
+  let chatId: string | undefined = undefined;
+
+  if (summary && summary.type === 'chatSummary') {
+    chatId = summary.chatId;
+    summaryText = `Below is the Chat Summary till now, this is chat summary before the conversation provided by the user 
+you should also use this as historical message while providing the response to the user.        
```

Comment:
We might think of introducing a limit how long the summary should be, wdyt?

Reply:
yes. lots of things I want to be configurable like the number of files in context buffer, summary size etc. but might be in a separate PR. otherwise it will be too big